packages/core/src/odata/common/request-builder/batch/batch-response-parser.ts

```diff
@@ -0,0 +1,376 @@
+ * @param response The string representation of a single response.
+ * @returns The parsed JSON representation of the response body.
+ */
+function parseResponseBody(response: string): Record<string, any> {
+  const responseBody = getResponseBody(response);
+  if (responseBody) {
+    try {
+      return JSON.parse(responseBody);
+    } catch (err) {
+      logger.error(
+        `Could not parse response body. Invalid JSON. Original Error: ${err}`
+      );
+    }
+  }
+  return {};
+}
+
+/**
+ * Parse a single response to the according response object type.
+ * @param response The string representation of a single response.
+ * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns The parsed response in the according response object representation.
+ */
+export function parseResponse(
+  response: string,
+  entityToConstructorMap: Record<string, Constructable<EntityBase>>,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+): WriteResponse | ReadResponse | ErrorResponse {
+  const responseData = {
+    body: parseResponseBody(response),
+    httpCode: parseHttpCode(response)
+  };
+
+  // retrieve response
+  if (responseData.httpCode === 200) {
+    return {
+      ...responseData,
+      type: getConstructor(
+        responseData.body,
+        entityToConstructorMap,
+        responseDataAccessor
+      ),
+      as: asReadResponse(responseData.body, responseDataAccessor, deserializer),
+      isSuccess: () => true
+    };
+  }
+
+  // change set sub response
+  if (responseData.httpCode === 201 || responseData.httpCode === 204) {
+    return {
+      ...responseData,
+      type: getConstructor(
+        responseData.body,
+        entityToConstructorMap,
+        responseDataAccessor
+      ),
+      as: asWriteResponse(responseData.body, responseDataAccessor, deserializer)
+    };
+  }
+
+  // error response
+  return { ...responseData, isSuccess: () => false };
+}
+
+/**
+ * Parse the complete batch HTTP response.
+ * @param batchResponse HTTP response.
+ * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns An array of parsed sub responses of the batch response.
+ */
+export function parseBatchResponse(
+  batchResponse: HttpResponse,
+  entityToConstructorMap: Record<string, Constructable<EntityBase>>,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+): (ErrorResponse | ReadResponse | WriteResponses)[] {
+  return splitBatchResponse(batchResponse).map(response => {
+    const contentType = getHeaderValue('content-type', parseHeaders(response));
+
+    // is change set response
+    if (isMultipartContentType(contentType)) {
```

Comment:
General: I do not like these if switches here `isMultipartContentType-> WriteResponse` and `isHttpContentType->ReadResponse | ErrorResponse` because in the `parseResponse` you have one more switch checks: `200-> ReadResponse` `201,204-> WriteResponse` and rest goes to Error. I would prefer three methos: `parseReapResponse`, `parseWriteResponse` and `parseErroroReposne` use these three methods in here and use a combination of status and contentType to select the right one.

Reply:
Yes, I agree. This was there before and I didn't have a good idea on how to improve it right away. I'll try to improve it and then discuss this with you.

Reply:
@FrankEssenberger Ok, so after our discussion yesterday I implemented the following changes: https://github.com/SAP/cloud-sdk/pull/471/commits/0cf2e5e7997d2d5db53849e965c3c7a4441b100e

I like the addition of the specific functions for parsing the responses. However, I don't like the early success check. As it makes the parsing asymmetric. The more I look at this the more I think that there are too many mixed concerns. Therefore I split the parsing and the transformation (and added a class for state) here: https://github.com/SAP/cloud-sdk/pull/471/commits/ca6ec83f5e2bdb08ba977fe2b113523e07dc07da

I still need to adjust tests, but I think this makes it much better.